src/base/lib/mapping/mapmach.c

```diff
@@ -0,0 +1,123 @@
+/*
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+
+/*
+ * Purpose: memory mapping library, Mach backend for macOS
+ *
+ * Authors: Bart Oldeman.
+ */
+
+#include <string.h>
+#include <sys/mman.h>
+
+#include <mach/mach_init.h>
+#include <mach/mach_vm.h>
+
+#include "dosemu_debug.h"
+#include "mapping.h"
+
+/* ------------------------------------------------------------ */
+
+static void *alias_mapping_mach(int cap, void *target, size_t mapsize, int protect, void *source)
+{
+  mach_vm_address_t targetaddr;
+  vm_prot_t cur, max;
+  int flags;
+
+  if (target != (void *)-1) {
+    flags = VM_FLAGS_FIXED | VM_FLAGS_OVERWRITE;
+    targetaddr = (vm_address_t)target;
+  } else {
+    flags = VM_FLAGS_ANYWHERE;
+    targetaddr = (vm_address_t)NULL;
+  }
+  if (mach_vm_remap(mach_task_self(), &targetaddr, mapsize, 0, flags,
+		    mach_task_self(), (mach_vm_address_t)source, FALSE, &cur, &max, VM_INHERIT_NONE)
+      != KERN_SUCCESS)
+    return MAP_FAILED;
+
+  target =  (void *)targetaddr;
+  mprotect(target, mapsize, protect);
+  return target;
+}
+
+static int open_mapping_mach(int cap)
+{
+  if (cap) Q_printf("MAPPING: open, cap=%s\n",
+		    decode_mapping_cap(cap));
+  return 1;
+}
+
+static void close_mapping_mach(int cap)
+{
+  Q_printf("MAPPING: close, cap=%s\n", decode_mapping_cap(cap));
+}
+
+static void *alloc_mapping_mach(int cap, size_t mapsize, void *target)
+{
+  int fixed = 0;
+
+  Q__printf("MAPPING: alloc, cap=%s, mapsize=%zx\n", cap, mapsize);
+  if (target != (void *)-1)
+    fixed = MAP_FIXED;
+  else
+    target = NULL;
+  return mmap(target, mapsize, PROT_READ | PROT_WRITE,
+    MAP_PRIVATE | MAP_ANONYMOUS | fixed, -1, 0);
```

Comment:
But does this work, by any chance,
with MAP_SHARED too? Then perhaps
it can be same as mapashm.

Reply:
I checked that it worked with MAP_SHARED as well.

So I can make it work with both Linux (mremap) and Mac (mach_vm_remap). That'll be for another day though..